pkg/restmapper/caching.go

```diff
@@ -17,124 +17,157 @@ import (
-		if !ok {
-			return nil
+		gv := gvr.GroupVersion()
+		cachedGV := c.cacheForGroupVersion(gv)
+
+		all, err := cachedGV.KindsFor(ctx, gvr, discovery)
+		if err != nil {
+			return nil, err
 		}
-		return []string{cachedgvr.toKind[gvr.Resource]}
+
+		matches = append(matches, all...)
+
+		return matches, nil
 	}
-	var kinds []string
 
-	for keyGVR, cachedgvr := range c.groupVersions {
-		if keyGVR.Group != gvr.Group {
+	allGroups, err := c.fetchAllGroups(ctx, discovery)
+	if err != nil {
+		return nil, err
+	}
+	for groupName, group := range allGroups {
+		if groupName != gvr.Group {
 			continue
 		}
-		kind, ok := cachedgvr.toKind[gvr.Resource]
-		if ok && kind != "" {
-			kinds = append(kinds, kind)
+		for _, version := range group.Versions {
+			gv := schema.GroupVersion{Group: groupName, Version: version.Version}
+			cachedGV := c.cacheForGroupVersion(gv)
+
+			all, err := cachedGV.KindsFor(ctx, gvr, discovery)
+			if err != nil {
+				return nil, err
+			}
+
+			matches = append(matches, all...)
 		}
 	}
-	return kinds
+	return matches, nil
 }
 
-// findRESTMapping returns the RESTMapping for the specified GVK, querying discovery if not cached.
-func (c *cache) findRESTMapping(ctx context.Context, discovery discovery.DiscoveryInterface, gv schema.GroupVersion, kind string) (*meta.RESTMapping, error) {
+func (c *cache) cacheForGroupVersion(gv schema.GroupVersion) *cachedGroupVersion {
 	c.mutex.Lock()
-	cached := c.groupVersions[gv]
+	cached := c.cachedGroupVersions[gv]
 	if cached == nil {
-		cached = &cachedGroupVersion{gv: gv, toKind: make(map[string]string)}
-		c.groupVersions[gv] = cached
+		cached = &cachedGroupVersion{gv: gv}
+		c.cachedGroupVersions[gv] = cached
 	}
 	c.mutex.Unlock()
-	return cached.findRESTMapping(ctx, discovery, kind)
+	return cached
+}
+
+// findRESTMapping returns the RESTMapping for the specified GVK, querying discovery if not cached.
+func (c *cache) findRESTMapping(ctx context.Context, discovery discovery.DiscoveryInterface, gv schema.GroupVersion, kind string) (*meta.RESTMapping, error) {
+	cachedGV := c.cacheForGroupVersion(gv)
+	return cachedGV.findRESTMapping(ctx, discovery, kind)
 }
 
 // findRESTMapping returns the RESTMapping for the specified GVK, querying discovery if not cached.
 func (c *cachedGroupVersion) findRESTMapping(ctx context.Context, discovery discovery.DiscoveryInterface, kind string) (*meta.RESTMapping, error) {
-	kinds, err := c.fetch(ctx, discovery)
+	resources, err := c.fetchServerResources(ctx, discovery)
 	if err != nil {
 		return nil, err
 	}
 
-	cached, found := kinds[kind]
-	if !found {
-		return nil, nil
+	for _, resource := range resources {
+		if resource.GVK().Kind == kind {
+			return resource.RESTMapping(), nil
+		}
 	}
-	return &meta.RESTMapping{
-		Resource:         c.gv.WithResource(cached.Resource),
-		GroupVersionKind: c.gv.WithKind(kind),
-		Scope:            cached.Scope,
-	}, nil
+	return nil, nil
 }
 
 // fetch returns the metadata, fetching it if not cached.
```

Comment:
```suggestion
// fetchServerResources returns the metadata, fetching it if not cached.
```
This method is renamed to `fetchServerResources`.
But it's small point for implementing.